File: Medium_Python/15_3Sum.py
class Solution:
    #先排序，然后遍历，left=i+1,right=len(nums)-1,若小于0，则加上left，若大于，则减去
    #如果当前值大于0，结束循环，如果当前下一个相等，跳过，如果sum=0时，左边+1相等，跳过，同理右边-1
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums)<3:
            return [];
        left=1
        right=len(nums)-1
        nums=sorted(nums)
        i=0
        returnarr=[]
        while i<len(nums)-2:
            if nums[i]>0:
                break
            while left<right:
                if nums[i]+nums[left]+nums[right]<0:
                    left=left+1
                elif nums[i]+nums[left]+nums[right]>0:
                    right=right-1
                else:
                    #if [nums[i],nums[left],nums[right]] not in returnarr:
                    returnarr.append([nums[i],nums[left],nums[right]])
                    while left< right and left+1<len(nums) and nums[left+1]==nums[left]:
                        left=left+1
                    while left<right and nums[right-1]==nums[right]:
                        right=right-1
                    left=left+1
            while i+1<len(nums) and nums[i+1]==nums[i]:
                i=i+1
            i=i+1
            left=i+1
            right=len(nums)-1
            
        return returnarr

File: Medium/15_3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        //数组长度小于3，直接返回
        //排序数组
        //若当前数字大于0，直接返回，若当前等于下一个数组，跳过 // left = i+1 right = n-1
        //如果等于0，移动左右指针，并确保跳过相同的
        //大于，则移动右指针，否则移动左边指针。
        vector<vector<int>> res;
        vector<int> temp;
        if(nums.size() < 3)
        {
            return res;
        }
        else
        {
            sort(nums.begin(),nums.end());
            int left, right;
            for(int i = 0; i < nums.size() - 2 ;i++)
            {
                if(nums[i] > 0)
                {
                    break;
                }
                left = i + 1;
                right = nums.size() - 1;
                while(left < right)
                {
                    if(nums[i] + nums[left] + nums[right] == 0)
                    {
                        temp.push_back(nums[i]);
                        temp.push_back(nums[left]);
                        temp.push_back(nums[right]);
                        res.push_back(temp);
                        temp.clear();    
                        
                        int cur = left;
                        while(cur+1 < nums.size()-1 && nums[left]==nums[cur+1])
                        {
                            cur++;
                        }
                        left = cur +1;
                    }
                    else if(nums[i] + nums[left] + nums[right] < 0)
                    {
                        left++;
                    }
                    else
                    {
                        right--;
                    }
                }    
                int current = i;
                while(current+1 < nums.size()-1 && nums[i]==nums[current+1])
                {
                    current++;
                }
                i = current;
            }     
        }
        return res;
    }
};
